File: raspidoor/sip/dialog.py

import time

from .request import *
from .auth import AuthenticationChallenge


class CallDeclinedError(Exception):
    pass


class Dialog:
    def __init__(self, transport, caller, username=None, password=None):
        self._transport = transport
        self._username = username
        self._password = password
        self._caller = caller
        self._cseq = 0
        self._call_id = "c{}".format(round(time.time() * 1000))
        self._transport.connect()

    def close(self):
        self._transport.close()

    def register(self):
        req = Request("REGISTER", "sip:{}".format(self._transport.host))
        req.header.\
            set("From", self._caller).\
            set("To", self._caller).\
            set("Contact", self._caller).\
            set("Max-Forwards", 70).\
            set("Expires", 7200)

        reponse = self._exchange(req)

    def invite(self, callee):
        req = Request("INVITE", callee)
        req.header.\
            set("From", self._caller).\
            set("To", callee).\
            set("Contact", self._caller).\
            set("Max-Forwards", 70)

        try:
            response = self._exchange(req)
            if response.status_code == RESPONSE_STATUS_OK:
                return self._bye(callee)
        except CallDeclinedError:
            return

    def _bye(self, callee):
        req = Request("BYE", callee)
        req.header.\
            set("From", self._caller).\
            set("To", callee).\
            set("Contact", self._caller).\
            set("Max-Forwards", 70)

        response = self._exchange(req)
        if response.status_code == RESPONSE_STATUS_OK:
            return
        else:
            raise ValueError("unexpected status code when sending BYE: {} {}".format(
                response.status_code, response.status_message))

    def _exchange(self, request):
        self._send(request)

        resp = self._transport.read_response()

        if resp.status_code == RESPONSE_STATUS_UNAUTHORIZED:
            wwwAuth = resp.header.get("www-authenticate")
            if wwwAuth == None:
                raise ValueError(
                    "Missing WWW-Authenticate header in response with status 401")

            authChallenge = AuthenticationChallenge.parse(wwwAuth)

            if not authChallenge.can_solve():
                raise ValueError(
                    "Unsolvable authentication challenge: {}", authChallenge.method)

            reqWithAuth = authChallenge.solve(
                request, self._username, self._password)
            self._send(request)

            resp = self._transport.read_response()

        while True:
            if resp.status_code > 399:
                raise self.handle_error_response(resp)
            elif resp.status_code < 200:
                resp = self._transport.read_response()
            else:
                break

        return resp

    def handle_error_response(self, response):
        if response.status_code == RESPONSE_STATUS_DECLINE:
            return CallDeclinedError()

        return ValueError("unexpected status code: {} {}".format(
            response.status_code, response.status_message))

    def _send(self, request):
        self._cseq += 1
        request.header.set_call_id(self._call_id).set_cseq(
            "{} {}".format(self._cseq, request.method))

        self._transport.send_request(request)

File: daemon/internal/sip/dialog.go
package sip

import (
	"errors"
	"fmt"
	"net/http"
	"strconv"
	"time"
)

const (
	StatusOK               = 200
	StatusRequestCancelled = 487
	StatusDecline          = 603
)

type Dialog struct {
	transport              Transport
	caller                 URI
	authenticationHandlers []AuthenticationHandler
	cseq                   int
	callID                 string
}

func NewDialog(transport Transport, caller URI, authenticationHandlers ...AuthenticationHandler) *Dialog {
	return &Dialog{
		transport:              transport,
		caller:                 caller,
		authenticationHandlers: authenticationHandlers,
		cseq:                   1,
		callID:                 fmt.Sprintf("c%d", time.Now().Unix()),
	}
}

func (d *Dialog) Ring(callee URI, maxRingingTime time.Duration) (bool, error) {
	inviteRequest := d.invite(callee, maxRingingTime)
	con, err := d.transport.Send(inviteRequest)
	if err != nil {
		return false, err
	}
	defer con.Close()

	inviteResponse, err := RecvFinal(con)
	if err != nil {
		return false, err
	}

	var authenticationChallenge AuthenticationChallenge

	if inviteResponse.StatusCode == http.StatusUnauthorized {
		authenticationChallenge, err = parseWWWAuthenticateHeader(inviteResponse.Header.Get("WWW-Authenticate"))
		if err != nil {
			return false, err
		}

		if err := d.authenticate(inviteRequest, authenticationChallenge); err != nil {
			return false, err
		}

		if err := con.Send(inviteRequest); err != nil {
			return false, err
		}
		inviteResponse, err = RecvFinal(con)
	}

	accepted := inviteResponse.StatusCode == StatusOK

	to := inviteResponse.Header.Get("To")

	ack, err := d.ack(callee, to, authenticationChallenge)
	if err != nil {
		return false, err
	}

	if err := con.Send(ack); err != nil {
		return false, err
	}

	d.cseq++

	bye := d.request("BYE", callee)
	bye.Header.Set("To", to)
	if err := con.Send(bye); err != nil {
		return false, err
	}

	resp, err := RecvFinal(con)
	if err != nil {
		return false, err
	}

	if resp.StatusCode != StatusOK {
		return false, fmt.Errorf("Got unexpected status from BYE: %d", resp.StatusCode)
	}

	return accepted, nil
}

func (d *Dialog) invite(callee URI, maxRingingTime time.Duration) *Request {
	r := d.request("INVITE", callee)
	r.Header.Add("Expires", strconv.Itoa(int(maxRingingTime.Seconds())))
	return r
}

func (d *Dialog) ack(callee URI, to string, authenticationChallenge AuthenticationChallenge) (*Request, error) {
	r := d.request("ACK", callee)
	r.Header.Set("To", to)
	r.SetBody("application/sdp", []byte(d.formatSDP()))

	if authenticationChallenge.Method != "" {
		err := d.authenticate(r, authenticationChallenge)
		if err != nil {
			return nil, err
		}
	}

	return r, nil
}

func (d *Dialog) formatSDP() string {
	sessionId := time.Now().UnixMilli()
	version := sessionId

	addr := "fe80::b644:fe20:c499:dd4e"

	return fmt.Sprintf("v=0\r\no=klingel1 %d %d IN IP6 %s\r\ns=\r\nc=IN IP6 %s\r\nt=0 0\r\n", sessionId, version, addr, addr)
}

func (d *Dialog) request(method string, callee URI) *Request {
	req := NewRequest(method, callee)
	req.Header.Set("From", d.caller.String())
	req.Header.Set("To", callee.String())
	req.Header.Set("Contact", d.caller.String())
	req.Header.Set("Max-Forwards", "70")
	req.Header.Set("Cseq", fmt.Sprintf("%d %s", d.cseq, req.Method))
	req.Header.Set("Call-ID", d.callID)

	return req
}

func (d *Dialog) authenticate(req *Request, c AuthenticationChallenge) error {
	for _, h := range d.authenticationHandlers {
		err := h.Solve(c, req)
		if err == nil {
			d.cseq++
			req.Header.Set("Cseq", fmt.Sprintf("%d %s", d.cseq, req.Method))
			return nil
		}

		if errors.Is(err, ErrUnsolveableAuthenticationChallenge) {
			continue
		}

		return err
	}

	return ErrUnsolveableAuthenticationChallenge
}
